File: Fundamentals_of_Software_Design/2_Algorithms/Сортировка миллиона целых чисел/main.py
import random
import time

# Размер массива
SIZE = 1000000

# Сортировка вставками
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

# Быстрая сортировка
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

# Поразрядная сортировка
def counting_sort_by_byte(arr, exp):
    count = [0] * 256
    output = [0] * len(arr)

    # Подсчет вхождений каждого байта
    for num in arr:
        index = (num // exp) % 256
        count[index] += 1

    # Изменяем count[i] так, чтобы оно содержало позиции элементов в output
    for i in range(1, 256):
        count[i] += count[i - 1]

    # Строим отсортированный массив
    for i in range(len(arr) - 1, -1, -1):
        num = arr[i]
        index = (num // exp) % 256
        output[count[index] - 1] = num
        count[index] -= 1

    # Копируем отсортированные элементы обратно в оригинальный массив
    for i in range(len(arr)):
        arr[i] = output[i]

def radix_sort(arr):
    # Проходим по каждому байту числа, начиная с младшего
    max_num = max(arr)
    exp = 1
    while max_num // exp > 0:
        counting_sort_by_byte(arr, exp)
        exp *= 256

# Генерация случайного массива
def generate_random_array(n):
    return [random.randint(0, 1000000) for _ in range(n)]

# Тестирование времени выполнения сортировок
def test_sorting_algorithms():
    arr1 = generate_random_array(SIZE)
    arr2 = arr1[:]
    arr3 = arr1[:]
    arr4 = arr1[:]

    # Сортировка вставками
    start = time.perf_counter()
    insertion_sort(arr1)
    end = time.perf_counter()
    print(f"Сортировка вставками: {end - start:.4f} секунд")

    # Быстрая сортировка
    start = time.perf_counter()
    arr2 = quick_sort(arr2)
    end = time.perf_counter()
    print(f"Быстрая сортировка: {end - start:.4f} секунд")

    # Поразрядная сортировка
    start = time.perf_counter()
    radix_sort(arr3)
    end = time.perf_counter()
    print(f"Поразрядная сортировка: {end - start:.4f} секунд")

    # Встроенная сортировка (sorted)
    start = time.perf_counter()
    arr4 = sorted(arr4)
    end = time.perf_counter()
    print(f"Встроенная сортировка (sorted): {end - start:.4f} секунд")

# Запуск тестов
test_sorting_algorithms()

File: Fundamentals_of_Software_Design/2_Algorithms/Сортировка миллиона целых чисел/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SIZE 1000000  // Размер массива

// Сортировка вставками
void insertion_sort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

// Быстрая сортировка
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return i + 1;
}

void quick_sort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quick_sort(arr, low, pi - 1);
        quick_sort(arr, pi + 1, high);
    }
}

// Поразрядная сортировка (по байтам)
void counting_sort_by_byte(int arr[], int n, int byte_index) {
    int count[256] = {0};
    int output[n];

    // Подсчитываем вхождения каждого байта
    for (int i = 0; i < n; i++) {
        int byte_value = (arr[i] >> (8 * byte_index)) & 0xFF;
        count[byte_value]++;
    }

    // Модифицируем массив count так, чтобы он содержал позиции элементов в output
    for (int i = 1; i < 256; i++) {
        count[i] += count[i - 1];
    }

    // Строим отсортированный массив
    for (int i = n - 1; i >= 0; i--) {
        int byte_value = (arr[i] >> (8 * byte_index)) & 0xFF;
        output[count[byte_value] - 1] = arr[i];
        count[byte_value]--;
    }

    // Копируем отсортированные элементы обратно в оригинальный массив
    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

void radix_sort(int arr[], int n) {
    for (int byte_index = 0; byte_index < 4; byte_index++) {  // Для 32-битных чисел 4 байта
        counting_sort_by_byte(arr, n, byte_index);
    }
}

// Функция для сравнения в qsort
int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

// Функция для генерации случайного массива
void generate_random_array(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % 1000000;
    }
}

int main() {
    srand(time(NULL));

    int arr1[SIZE], arr2[SIZE], arr3[SIZE], arr4[SIZE];
    
    // Генерируем случайные данные для тестов
    generate_random_array(arr1, SIZE);
    for (int i = 0; i < SIZE; i++) arr2[i] = arr1[i];
    for (int i = 0; i < SIZE; i++) arr3[i] = arr1[i];
    for (int i = 0; i < SIZE; i++) arr4[i] = arr1[i];

    clock_t start, end;

    // Тестируем сортировку вставками
    start = clock();
    insertion_sort(arr1, SIZE);
    end = clock();
    printf("Сортировка вставками: %.2f секунд\n", (double)(end - start) / CLOCKS_PER_SEC);

    // Тестируем быструю сортировку
    start = clock();
    quick_sort(arr2, 0, SIZE - 1);
    end = clock();
    printf("Быстрая сортировка: %.2f секунд\n", (double)(end - start) / CLOCKS_PER_SEC);

    // Тестируем поразрядную сортировку
    start = clock();
    radix_sort(arr3, SIZE);
    end = clock();
    printf("Поразрядная сортировка: %.2f секунд\n", (double)(end - start) / CLOCKS_PER_SEC);

    // Тестируем библиотечную qsort
    start = clock();
    qsort(arr4, SIZE, sizeof(int), compare);
    end = clock();
    printf("Библиотечная qsort: %.2f секунд\n", (double)(end - start) / CLOCKS_PER_SEC);

    return 0;
}
